### PR TITLE
Manage tokengroups and assign them to tokens

### PR DESCRIPTION
**Pull request by cornelinux**
Add the tokengroup table for defining token groups and the table "tokentokengroup" for assigning tokens to tokengroups.

Working on #3299

### CODE REVIEW

File: privacyidea/static/components/config/views/config.tokengroup.html [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+                            translate>List Tokengroup definitions</a></li>
+                <li role="presentation"
+                    ng-class="{ active: $state.includes('config.tokengroup.add')}"
+                    ng-show="checkRight('tokengroup_add')">
+                    <a ui-sref="config.tokengroup.add" translate>
+                        New Tokengroup definition</a></li>
+                <hr>
+                <li role="presentation">
+                    <a target="_external" ng-href="{{ 'https://privacyidea.readthedocs.io/en/v'+
+                       privacyideaVersionNumber+'/configuration/tokengroup.html'}}">
```
Comment (NNLemling) [inline privacyidea/static/components/config/views/config.tokengroup.html:17]: please rename this to tokengroups.html

File: privacyidea/api/token.py [hunk 2]
```diff
@@ -1225,3 +1226,58 @@ def delete_tokeninfo_api(serial, key):
...
+def assign_tokengroup_api(serial, groupname=None):
+    """
+    Assigns a token to a given tokengroup.
+
+    If no groupname is given, we expect a body data "groups", that
+    contains a list of tokengroups. tokengroups that are
+    not contained in this list, will be removed.
+
+    :jsonparam basestring serial: the serial number of the token
+    :jsonparam basestring groupname: The name of the tokengroup
```
Comment (NNLemling) [inline privacyidea/api/token.py:1246]: please describe parameter "groups".
Which is used in line 1255.

File: privacyidea/lib/tokenclass.py [hunk 3]
```diff
@@ -245,6 +245,22 @@ def add_user(self, user, report=None):
         # set the tokenrealm
         self.set_realms([user.realm])
 
+    def add_tokengroup(self, tokengroup=None, tokengroup_id=None):
+        """
+        Adds a new tokengroup to this token.
```
Comment (NNLemling) [inline privacyidea/lib/tokenclass.py:250]: add new line between parameter and description.

File: privacyidea/lib/tokenclass.py [hunk 4]
```diff
@@ -245,6 +245,22 @@ def add_user(self, user, report=None):
...
+        :param tokengroup: The name of the token group to add
+        :type tokengroup: basestring
+        :param tokengroup_id: The id of the tokengroup to add
+        :type tokengroup_id: int
+        :return: True
+        """
+        if not tokengroup and not tokengroup_id:
+            raise ParameterError("You either need to specify a tokengroup name or id.")
+        r = TokenTokengroup(token_id=self.token.id,
+                           tokengroup_id=tokengroup_id,
```
Comment (NNLemling) [inline privacyidea/lib/tokenclass.py:260]: add whitespaceses in line 260, 261.

File: privacyidea/models.py [hunk 5]
```diff
@@ -3080,3 +3139,107 @@ def __init__(self, timestamp, key, value):
...
+        self.name = groupname
+        self.Description = description
+
+    def delete(self):
+        ret = self.id
+        # delete all TokenTokenGroup
+        db.session.query(TokenTokengroup)\
+                  .filter(TokenTokengroup.tokengroup_id == ret)\
+                  .delete()
+        # delete the realm
```
Comment (NNLemling) [inline privacyidea/models.py:3167]: rename to "realm" to "tokengroup".

**Review by NNLemling — COMMENTED** [review]
Please fix the suggested changes

File: privacyidea/models.py (hunk 5 above)
Comment (cornelinux) [reply privacyidea/models.py]: done in cfc6b42a

File: privacyidea/api/token.py (hunk 2 above)
Comment (cornelinux) [reply privacyidea/api/token.py:1246]: done in cfc6b42a

File: privacyidea/static/components/config/views/config.tokengroup.html (hunk 1 above)
Comment (cornelinux) [reply privacyidea/static/components/config/views/config.tokengroup.html]: done in cfc6b42a

**Pull request merged**